File: Excel/585 Challenge.py
import pandas as pd

path = "585 List the Factorials.xlsx"
input = pd.read_excel(path, usecols="A", skiprows=1, nrows=10)
test = pd.read_excel(path, usecols="B:C", skiprows=1, nrows=5)

def is_factorial(n):
    if n < 1:
        return False
    factorial = 1
    i = 1
    while factorial < n:
        i += 1
        factorial *= i
    return factorial == n

def factorial_of(n):
    if n < 1:
        return None
    factorial = 1
    i = 1
    while factorial < n:
        i += 1
        factorial *= i
    return i if factorial == n else None

input['is_fact'] = input.iloc[:, 0].apply(is_factorial)
result = input[input['is_fact']].copy()
result['Factorial Of'] = result.iloc[:, 0].apply(factorial_of)
result = result.drop(columns=['is_fact']).reset_index(drop=True)
result.columns = ['Number', 'Factorial Of']

print(result.equals(test)) # True

File: Excel/500-599/585/585 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/585 List the Factorials.xlsx"
input = read_excel(path, range = "A2:A11")
test  = read_excel(path, range = "B2:C7")


is_factorial <- function(n) {
  if (n < 1)
    return(FALSE)
  factorial <- 1
  i <- 1
  while (factorial < n) {
    i <- i + 1
    factorial <- factorial * i
  }
  return(factorial == n)
}

factorial_of <- function(n) {
  if (n < 1)
    return(NA)
  factorial <- 1
  i <- 1
  while (factorial < n) {
    i <- i + 1
    factorial <- factorial * i
  }
  return(ifelse(factorial == n, i, NA))
}

result = input %>%
  mutate(is_fact = map_lgl(Numbers, is_factorial)) %>%
  filter(is_fact) %>%
  mutate(`Factorial Of` = map_dbl(Numbers, factorial_of)) %>%
  select(-is_fact)

all.equal(result, test, check.attributes = FALSE)
#> [1] TRUE